### DISCUSSION TITLE
Fix non unit normals

### DISCUSSION THREAD

**Discussion by looeee** [donmccurdy/glTF-Transform#263] [Q&A] [answered]
A common error that I'm encountering is `Vector3 at accessor indices 102..104 is not of unit length: 0`. 

Is within scope to add a function to normalize normals? Or even to just do it automatically on read/write?

**Comment by lexaknyazev** [comment]
Zero-length normal vector usually indicates some authoring issue - there's probably not enough information in the asset to generate it automatically.

@donmccurdy WDYT?

**Comment by donmccurdy** [comment, marked as answer]
*(moved to discussion)*

@looeee do you know if the 0-length normals are present in the original models, or coming from some step in the optimization processing? E.g. if the `quantize()` function is producing bad normals I'd certainly want to fix that.

Normalizing normals or tangents is not too hard:

```ts
import { length, divide } from 'gl-matrix/vec3';

for (const mesh of document.getRoot().listMeshes()) {
  mesh.listPrimitives().forEach(normalize);
}

function normalize(primitive: Primitive) {
  const normals = primitive.getAttribute('NORMAL');

  const v = [0, 0, 0];
  const scale = [1, 1, 1];

  for (let i = 0, il = normals.getCount(); i < il; i++) {
    normals.getElement(i, v);
    divide(v, v, scale.fill(length(v) || 1));
    normals.setElement(i, v);
  }
}
```

^Same thing for tangents, but ignoring the `.w` component which must be +/- 1. 

But as @lexaknyazev mentions, 0-length normals can't simply be normalized, we'd be guessing without knowing whether the normals are meant to be smooth, hard, or customized, and the triangle could be degenerate anyway. I suspect that many of the Google Blocks models do not need vertex normals at all — they all use a low-poly flat shading effect, I think — if that's what you're optimizing here?

**Comment by looeee** [comment]
Ah, I didn't realize the "0" at the end of the error was stating the length. Yeah, that complicates things. 
These particular models were ones that glTF-transform couldn't handle originally so I ran them through gltfpack. 

I'll try and track down at what point the error is generated.
